File: src/main/java/com/example/finsentinel/service/trading/engine/AlpacaTradingEngine.java
package com.example.finsentinel.service.trading.engine;

import lombok.extern.slf4j.Slf4j;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.ObjectMapper;
import tools.jackson.databind.json.JsonMapper;

import java.io.IOException;
import java.math.BigDecimal;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Alpaca Markets trading engine for US equities.
 *
 * <p>Communicates with the Alpaca REST API v2 using Java 21's built-in {@link HttpClient}.
 * Not a Spring bean — instantiated by the {@code TradingEngineFactory}.
 *
 * <p>Default base URL targets the Alpaca paper-trading sandbox.
 */
@Slf4j
public class AlpacaTradingEngine implements TradingEngine {

    private static final String DEFAULT_BASE_URL = "https://paper-api.alpaca.markets";
    private static final ObjectMapper objectMapper = JsonMapper.builder().build();

    private final String apiKey;
    private final String secretKey;
    private final String baseUrl;
    private final HttpClient httpClient;

    public AlpacaTradingEngine(String apiKey, String secretKey, String baseUrl) {
        this.apiKey = apiKey;
        this.secretKey = secretKey;
        this.baseUrl = (baseUrl != null) ? baseUrl : DEFAULT_BASE_URL;
        this.httpClient = HttpClient.newHttpClient();
    }

    // ──────────────────────────────── TradingEngine contract ────────────────────

    @Override
    public String engineName() {
        return "alpaca";
    }

    @Override
    public OrderResult placeOrder(OrderRequest request) {
        try {
            Map<String, Object> body = buildOrderBody(request);
            String json = objectMapper.writeValueAsString(body);

            HttpResponse<String> response = sendPost("/v2/orders", json);
            JsonNode node = objectMapper.readTree(response.body());

            return parseOrderNode(node);
        } catch (Exception e) {
            log.error("Alpaca placeOrder failed for {}: {}", request.symbol(), e.getMessage(), e);
            String errorMsg = e.getMessage() != null ? e.getMessage() : e.getClass().getSimpleName();
            return new OrderResult(false, null, "rejected", BigDecimal.ZERO, BigDecimal.ZERO,
                    errorMsg, null);
        }
    }

    @Override
    public List<PositionInfo> getPositions() {
        try {
            HttpResponse<String> response = sendGet("/v2/positions");
            JsonNode array = objectMapper.readTree(response.body());

            List<PositionInfo> positions = new ArrayList<>();
            for (JsonNode node : array) {
                positions.add(new PositionInfo(
                        node.get("symbol").asText(),
                        node.get("side").asText(),
                        new BigDecimal(node.get("qty").asText()),
                        new BigDecimal(node.get("avg_entry_price").asText()),
                        new BigDecimal(node.get("current_price").asText()),
                        new BigDecimal(node.get("market_value").asText()),
                        new BigDecimal(node.get("unrealized_pl").asText()),
                        new BigDecimal(node.get("cost_basis").asText())
                ));
            }
            return positions;
        } catch (Exception e) {
            log.error("Alpaca getPositions failed: {}", e.getMessage(), e);
            return List.of();
        }
    }

    @Override
    public List<OrderResult> getOrders() {
        try {
            HttpResponse<String> response = sendGet("/v2/orders?status=all&limit=50");
            JsonNode array = objectMapper.readTree(response.body());

            List<OrderResult> orders = new ArrayList<>();
            for (JsonNode node : array) {
                orders.add(parseOrderNode(node));
            }
            return orders;
        } catch (Exception e) {
            log.error("Alpaca getOrders failed: {}", e.getMessage(), e);
            return List.of();
        }
    }

    @Override
    public List<OrderResult> syncOrders() {
        try {
            HttpResponse<String> response = sendGet("/v2/orders?status=open&limit=50");
            JsonNode array = objectMapper.readTree(response.body());
            List<OrderResult> orders = new ArrayList<>();
            for (JsonNode node : array) {
                orders.add(parseOrderNode(node));
            }
            return orders;
        } catch (Exception e) {
            log.error("Alpaca syncOrders failed: {}", e.getMessage(), e);
            return List.of();
        }
    }

    @Override
    public AccountInfo getAccount() {
        try {
            HttpResponse<String> response = sendGet("/v2/account");
            JsonNode node = objectMapper.readTree(response.body());

            return new AccountInfo(
                    new BigDecimal(node.get("cash").asText()),
                    new BigDecimal(node.get("portfolio_value").asText()),
                    new BigDecimal(node.get("equity").asText()),
                    new BigDecimal(node.get("buying_power").asText()),
                    node.has("unrealized_pl") && !node.get("unrealized_pl").isNull()
                            ? new BigDecimal(node.get("unrealized_pl").asText()) : BigDecimal.ZERO,
                    node.has("realized_pl") && !node.get("realized_pl").isNull()
                            ? new BigDecimal(node.get("realized_pl").asText()) : BigDecimal.ZERO
            );
        } catch (Exception e) {
            log.error("Alpaca getAccount failed: {}", e.getMessage(), e);
            return new AccountInfo(
                    BigDecimal.ZERO, BigDecimal.ZERO, BigDecimal.ZERO,
                    BigDecimal.ZERO, BigDecimal.ZERO, BigDecimal.ZERO
            );
        }
    }

    @Override
    public boolean cancelOrder(String orderId) {
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + "/v2/orders/" + orderId))
                    .header("APCA-API-KEY-ID", apiKey)
                    .header("APCA-API-SECRET-KEY", secretKey)
                    .DELETE()
                    .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            checkStatus(response);
            return true;
        } catch (Exception e) {
            log.error("Alpaca cancelOrder failed for {}: {}", orderId, e.getMessage(), e);
            return false;
        }
    }

    // ──────────────────────────────── Order parsing ────────────────────────────

    private OrderResult parseOrderNode(JsonNode node) {
        return new OrderResult(
                true,
                node.get("id").asText(),
                mapOrderStatus(node.get("status").asText()),
                node.has("filled_avg_price") && !node.get("filled_avg_price").isNull()
                        ? new BigDecimal(node.get("filled_avg_price").asText()) : BigDecimal.ZERO,
                node.has("filled_qty") && !node.get("filled_qty").isNull()
                        ? new BigDecimal(node.get("filled_qty").asText()) : BigDecimal.ZERO,
                null,
                node.has("filled_at") && !node.get("filled_at").isNull()
                        ? LocalDateTime.parse(node.get("filled_at").asText().substring(0, 19)) : null
        );
    }

    // ──────────────────────────────── HTTP helpers ──────────────────────────────

    private HttpResponse<String> sendGet(String path) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + path))
                .header("APCA-API-KEY-ID", apiKey)
                .header("APCA-API-SECRET-KEY", secretKey)
                .GET()
                .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        checkStatus(response);
        return response;
    }

    private HttpResponse<String> sendPost(String path, String jsonBody) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + path))
                .header("APCA-API-KEY-ID", apiKey)
                .header("APCA-API-SECRET-KEY", secretKey)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
                .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        checkStatus(response);
        return response;
    }

    private void checkStatus(HttpResponse<String> response) throws IOException {
        if (response.statusCode() >= 400) {
            throw new IOException("Alpaca API error " + response.statusCode() + ": " + response.body());
        }
    }

    // ──────────────────────────────── Mapping helpers ───────────────────────────

    /**
     * Builds the JSON body for the Alpaca POST /v2/orders endpoint.
     */
    private Map<String, Object> buildOrderBody(OrderRequest req) {
        Map<String, Object> body = new HashMap<>();
        body.put("symbol", req.symbol());
        body.put("side", req.side());
        body.put("type", req.type());
        body.put("time_in_force", req.timeInForce() != null ? req.timeInForce() : "day");

        if (req.qty() != null) {
            body.put("qty", req.qty().toPlainString());
        }
        if (req.notional() != null) {
            body.put("notional", req.notional().toPlainString());
        }
        if (req.price() != null) {
            body.put("limit_price", req.price().toPlainString());
        }
        if (req.stopPrice() != null) {
            body.put("stop_price", req.stopPrice().toPlainString());
        }

        return body;
    }

    /**
     * Maps Alpaca order status strings to the standardized FinSentinel status values.
     *
     * <ul>
     *   <li>{@code filled} -> {@code filled}</li>
     *   <li>{@code canceled} -> {@code cancelled}</li>
     *   <li>{@code rejected}, {@code expired} -> {@code rejected}</li>
     *   <li>everything else -> {@code pending}</li>
     * </ul>
     */
    private String mapOrderStatus(String alpacaStatus) {
        return switch (alpacaStatus) {
            case "filled" -> "filled";
            case "canceled" -> "cancelled";
            case "rejected", "expired" -> "rejected";
            default -> "pending";
        };
    }
}

File: apps/api/src/trading/engines/alpaca-trading.engine.ts
import type { TradingEngine } from '../interfaces/trading-engine';
import type {
  OrderRequest,
  OrderResult,
  PositionInfo,
  AccountInfo,
  MarketClock,
} from '../interfaces/types';

/**
 * Alpaca Markets trading engine for US equities.
 *
 * Communicates with the Alpaca REST API v2 using native `fetch()`.
 * NOT an Injectable -- instantiated by BrokerRegistry at runtime.
 *
 * Default base URL targets the Alpaca paper-trading sandbox.
 *
 * Implements the Alpaca-backed trading engine.
 */
export class AlpacaTradingEngine implements TradingEngine {
  private static readonly DEFAULT_BASE_URL = 'https://paper-api.alpaca.markets';

  private readonly baseUrl: string;

  constructor(
    private readonly apiKey: string,
    private readonly secretKey: string,
    baseUrl?: string,
  ) {
    this.baseUrl = baseUrl ?? AlpacaTradingEngine.DEFAULT_BASE_URL;
  }

  // ---------------------------------------------------------------------------
  // TradingEngine interface
  // ---------------------------------------------------------------------------

  engineName(): string {
    return 'alpaca';
  }

  async placeOrder(request: OrderRequest): Promise<OrderResult> {
    try {
      const body = this.buildOrderBody(request);
      const response = await this.sendPost('/v2/orders', body);
      const node = await response.json();

      return this.parseOrderNode(node);
    } catch (err: unknown) {
      const errorMessage = err instanceof Error ? err.message : String(err);
      return {
        success: false,
        orderId: '',
        status: 'rejected',
        filledQty: '0',
        avgPrice: '0',
        errorMessage,
        timestamp: new Date().toISOString(),
      };
    }
  }

  async getPositions(): Promise<PositionInfo[]> {
    try {
      const response = await this.sendGet('/v2/positions');
      const array = (await response.json()) as Record<string, string>[];

      return array.map((node) => ({
        symbol: node['symbol']!,
        side: node['side'] as 'long' | 'short',
        qty: node['qty']!,
        avgEntryPrice: node['avg_entry_price'],
        avgCost: node['avg_entry_price']!,
        currentPrice: node['current_price']!,
        marketValue: node['market_value'],
        unrealizedPnL: node['unrealized_pl']!,
        costBasis: node['cost_basis'],
      }));
    } catch (err: unknown) {
      // Return an empty result set on fetch failure.
      return [];
    }
  }

  async getOrders(): Promise<OrderResult[]> {
    try {
      const response = await this.sendGet('/v2/orders?status=all&limit=50');
      const array = (await response.json()) as Record<string, unknown>[];

      return array.map((node) => this.parseOrderNode(node));
    } catch (err: unknown) {
      return [];
    }
  }

  async syncOrders(): Promise<OrderResult[]> {
    try {
      const response = await this.sendGet('/v2/orders?status=open&limit=50');
      const array = (await response.json()) as Record<string, unknown>[];

      return array.map((node) => this.parseOrderNode(node));
    } catch (err: unknown) {
      return [];
    }
  }

  async getMarketClock(): Promise<MarketClock> {
    try {
      const response = await this.sendGet('/v2/clock');
      const node = (await response.json()) as Record<string, unknown>;

      return {
        isOpen: node['is_open'] as boolean,
        nextOpen: node['next_open'] as string,
        nextClose: node['next_close'] as string,
        timestamp: node['timestamp'] as string,
      };
    } catch (err: unknown) {
      return {
        isOpen: false,
        nextOpen: null,
        nextClose: null,
        timestamp: new Date().toISOString(),
      };
    }
  }

  async getAccount(): Promise<AccountInfo> {
    try {
      const response = await this.sendGet('/v2/account');
      const node = (await response.json()) as Record<string, string>;

      const cash = node['cash'] ?? '0';
      const portfolioValue = node['portfolio_value'] ?? '0';
      const equity = node['equity'] ?? '0';
      const buyingPower = node['buying_power'] ?? '0';
      const unrealizedPnL = node['unrealized_pl'] ?? '0';
      const realizedPnL = node['realized_pl'] ?? '0';

      return {
        totalValue: Number(portfolioValue),
        cashValue: Number(cash),
        buyingPower: Number(buyingPower),
        cash,
        portfolioValue,
        equity,
        unrealizedPnL,
        realizedPnL,
      };
    } catch (err: unknown) {
      return {
        totalValue: 0,
        cashValue: 0,
        buyingPower: 0,
      };
    }
  }

  async cancelOrder(orderId: string): Promise<boolean> {
    try {
      await this.sendDelete(`/v2/orders/${orderId}`);
      return true;
    } catch (err: unknown) {
      return false;
    }
  }

  // ---------------------------------------------------------------------------
  // Order parsing
  // ---------------------------------------------------------------------------

  private parseOrderNode(node: Record<string, unknown>): OrderResult {
    const filledAvgPrice =
      node['filled_avg_price'] != null ? String(node['filled_avg_price']) : '0';
    const filledQty = node['filled_qty'] != null ? String(node['filled_qty']) : '0';
    const filledAt = node['filled_at'] != null ? String(node['filled_at']) : undefined;

    return {
      success: true,
      orderId: String(node['id'] ?? ''),
      symbol: node['symbol'] as string | undefined,
      side: node['side'] as 'buy' | 'sell' | undefined,
      status: this.mapOrderStatus(String(node['status'] ?? '')),
      filledPrice: filledAvgPrice,
      filledQty,
      avgPrice: filledAvgPrice,
      errorMessage: null,
      timestamp: filledAt ?? new Date().toISOString(),
    };
  }

  // ---------------------------------------------------------------------------
  // HTTP helpers
  // ---------------------------------------------------------------------------

  private authHeaders(): Record<string, string> {
    return {
      'APCA-API-KEY-ID': this.apiKey,
      'APCA-API-SECRET-KEY': this.secretKey,
    };
  }

  private async sendGet(path: string): Promise<Response> {
    const response = await fetch(`${this.baseUrl}${path}`, {
      method: 'GET',
      headers: this.authHeaders(),
    });
    this.checkStatus(response);
    return response;
  }

  private async sendPost(path: string, body: Record<string, unknown>): Promise<Response> {
    const response = await fetch(`${this.baseUrl}${path}`, {
      method: 'POST',
      headers: {
        ...this.authHeaders(),
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
    this.checkStatus(response);
    return response;
  }

  private async sendDelete(path: string): Promise<Response> {
    const response = await fetch(`${this.baseUrl}${path}`, {
      method: 'DELETE',
      headers: this.authHeaders(),
    });
    this.checkStatus(response);
    return response;
  }

  private checkStatus(response: Response): void {
    if (!response.ok) {
      throw new Error(`Alpaca API error ${response.status}: ${response.statusText}`);
    }
  }

  // ---------------------------------------------------------------------------
  // Mapping helpers
  // ---------------------------------------------------------------------------

  /**
   * Builds the JSON body for the Alpaca POST /v2/orders endpoint.
   */
  private buildOrderBody(req: OrderRequest): Record<string, unknown> {
    const body: Record<string, unknown> = {
      symbol: req.symbol,
      side: req.side,
      type: req.type,
      time_in_force: req.timeInForce ?? 'day',
    };

    if (req.qty != null) {
      body['qty'] = req.qty;
    }
    if (req.notional != null) {
      body['notional'] = req.notional;
    }
    if (req.price != null) {
      body['limit_price'] = req.price;
    }
    if (req.stopPrice != null) {
      body['stop_price'] = req.stopPrice;
    }

    return body;
  }

  /**
   * Maps Alpaca order status strings to standardized FinSentinel status values.
   *
   * - "filled" -> "filled"
   * - "canceled" -> "cancelled"
   * - "rejected" | "expired" -> "rejected"
   * - everything else -> "pending"
   */
  private mapOrderStatus(alpacaStatus: string): string {
    switch (alpacaStatus) {
      case 'filled':
        return 'filled';
      case 'canceled':
        return 'cancelled';
      case 'rejected':
      case 'expired':
        return 'rejected';
      default:
        return 'pending';
    }
  }
}
